### PR TITLE
Redirect www.code.org to code.org

### PR DESCRIPTION
**Pull request by cat5inthecradle**
This creates a new cloudfront distribution, using a lambda@edge function to issue a redirect for all traffice to `www.code.org` to `code.org` - and likewise for lower environments.

The redirect is a 302. It would probably be reasonable to make it a 301, but a 301 is harder to change away from in the future, so this is a safer choice. It could be converted to a 301 at any time in the future.

## Links

- jira ticket: [INF-506](https://codedotorg.atlassian.net/browse/INF-506)

## Testing story

With the exception of the Realtime Log config, this has been tested in a sandbox, using this template:

https://gist.github.com/cat5inthecradle/ecb94d418f7472679f0d14b77212a6bf

This was deployed to codeorg-dev aws, and may still be viewable at https://darin.dev-code.org/about/values

## Deployment strategy

This can be deployed safely at any time, as it only adds new resources that do not impact the current infrastructure. There is a follow up task to update the `www.code.org` DNS record to point to the new cloudfront distribution.

## Follow-up work

* Continue Jira Ticket: https://codedotorg.atlassian.net/browse/INF-506

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/cdo/aws/cloudfront.rb [hunk 1]
```diff
@@ -34,7 +34,7 @@ class CloudFront
       pegasus: {
         # NOTE: Keep this list in sync with the call to AWS::CloudFront.distribution_config in cloud_formation_stack.yml.erb.
         # CloudFormation stack should be refactored to reference this configuration in the future.
-        aliases: [CDO.pegasus_hostname, CDO.advocacy_hostname] + CDO.partners.map {|x| CDO.canonical_hostname("#{x}.code.org")},
+        aliases: [CDO.pegasus_hostname, CDO.pegasus_www_hostname, CDO.advocacy_hostname] + CDO.partners.map {|x| CDO.canonical_hostname("#{x}.code.org")},
```
Comment (sureshc) [inline lib/cdo/aws/cloudfront.rb]: I think we don’t use the partners anymore (`CDO.partners` is empty on all of our environments, I think). Perhaps this is an opportunity to clean up that complexity?

Comment (cat5inthecradle) [reply lib/cdo/aws/cloudfront.rb]: That's what I was thinking. Do you know where that gets populated?

**Comment by sureshc** [comment]
Should we add an integration test or user interface test that exercises the `www.` domain name?

**Comment by Hamms** [comment]
For context, why _do_ we want to support this subdomain?

**Comment by sureshc** [comment]
> For context, why _do_ we want to support this subdomain?

There are users navigating directly to `www.code.org` currently according to @cat5inthecradle analysis of HTTP logs.

**Comment by Hamms** [comment]
I don't know that adding such a high-level record is the right solution for invalid user activity like that. I get that there are historical reasons why the `www` subdomain was once treated specially, but unless we are doing something with that particular subdomain I really don't like the idea of hardcoding it like this.

Is it feasible to put in some logic somewhere that generically handles any otherwise-undefined subdomain that people try to navigate directly to?

File: lib/cdo/aws/cloudfront.rb (hunk 1 above)
Comment (cat5inthecradle) [reply lib/cdo/aws/cloudfront.rb]: if it's empty everywhere (checking the sources listed in config/README) then lets get rid of it.

**Comment by cat5inthecradle** [comment]
> Is it feasible to put in some logic somewhere that generically handles any otherwise-undefined subdomain that people try to navigate directly to?

I would agree with this - but I wonder if `www.code.org` is the one exception.

In the past I've used a small nginx/apache proxy server to issue 301's, but that was mostly for website transitions, not long term.

**Comment by cat5inthecradle** [comment]
@Hamms  - I talked to @sureshc about this a little bit this afternoon.

I'd like to ultimately see a 301 redirect from `www.code.org` to `code.org`. There are a few ways to do that, and one of them is cloudfront and lambda@edge. Basically a small proxy function at cloudfront edge nodes that would issue the 301. It would make this PR obsolete, because we can isolate the logic into it's own (smaller cheaper) cloudfront distribution. We'd then ensure that A) everybody gets pointed to `code.org` and also that B) other environments deployed with this stack get their own `www` redirect (`www.staging` >> `staging`).

One of the [example use-cases](https://aws.amazon.com/blogs/networking-and-content-delivery/handling-redirectsedge-part1/) for this lambda@edge functionality is implementing a tinyurl-like service, which is pretty close to what we want.

I will probably close this PR after tomorrow's stand-up.

**Comment by Hamms** [comment]
That sounds fantastic! A redirect in general seems like the best way to handle somewhat-expected-but-still-invalid urls, and at least to me feels a lot better for this specific use case.

File: aws/cloudformation/cloud_formation_stack.yml.erb [hunk 2]
```diff
@@ -254,9 +256,15 @@ Resources:
           ResourceRecords: [!GetAtt <%=daemon%>.PublicIp]
 <%   end -%>
 <% end -%>
+
+<%  if environment == :production -%>
```
Comment (sureshc) [inline aws/cloudformation/cloud_formation_stack.yml.erb]: I think it’s worth adding the `www.` redirect logic to all environments that use CloudFront (`cdn_enabled`). That would enable to us to potentially add an integration or browser test that exercises the `www.` redirect as well. I think we’d insert that logic [in this block](https://github.com/code-dot-org/code-dot-org/blob/a00c195a1ec1205af09692d8cad9b602f5e7bd65/aws/cloudformation/cloud_formation_stack.yml.erb#L211-L240)?

File: aws/cloudformation/components/www_redirect.yml.erb [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        PriceClass: PriceClass_All
+        Aliases:
+          - www.code.org
+        Origins:
+          # An origin is required, but will never receive traffic
+          - Id: PrimaryOrigin
+            DomainName: code.org
+            CustomOriginConfig:
+              OriginProtocolPolicy: match-viewer
+        DefaultCacheBehavior:
```
Comment (sureshc) [inline aws/cloudformation/components/www_redirect.yml.erb:16]: Would be useful to set `RealtimeLogConfigArn`  To the imported `AccessLogs-Config` so we’re logging the HTTP requests before they get redirected.

File: aws/cloudformation/cloud_formation_stack.yml.erb (hunk 2 above)
Comment (cat5inthecradle) [reply aws/cloudformation/cloud_formation_stack.yml.erb]: plus then we get to test outside of prod!

**Review by sureshc — APPROVED** [review]
We're tracking a separate Task to add a user interface test that exercises this logic.

**Pull request merged**